### PR TITLE
forge script: deploy upgradable NFTAccountant

### PR DESCRIPTION
**Pull request by anna-carroll**
## Motivation
As part of reboot process, need to deploy & configure an upgradable NFTAccountant

## Solution
Implement forge script to do so

## PR Checklist

- [ ] Added Tests
- [ ] Updated Documentation
- [x] Updated CHANGELOG.md for the appropriate package

### CODE REVIEW

File: packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+contract DeployAccountant is Script, Config {
+    // entrypoint
+    function deploy(string calldata configFile, string calldata domain) public {
+        __Config_initialize(configFile);
+        deployAccountant(domain);
+    }
+
+    // Deploys & configures the NFTAccountant with upgrade setup
+    function deployAccountant(string calldata domain) internal {
+        address fundsRecipient = vm.addr(567); // TODO
```
Comment (anna-carroll) [inline packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol:21]: we conceptualized of the config as a place to store all of the information & configuration necessary to deploy & operate the Nomad protocol. This information is necessary to deploy the protocol. I'd like to add it to the config.

File: packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        address accountantOwner = vm.addr(89); // TODO
```
Comment (anna-carroll) [inline packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol:22]: we conceptualized of the config as a place to store all of the information & configuration necessary to deploy & operate the Nomad protocol. This information is necessary to deploy the protocol. I'd like to add it to the config.

File: packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import {Config} from "../Config.sol";
+
+import {AllowListNFTRecoveryAccountant} from "@nomad-xyz/contracts-bridge/contracts/accountants/NFTAccountant.sol";
+import {UpgradeBeacon} from "@nomad-xyz/contracts-core/contracts/upgrade/UpgradeBeacon.sol";
+import {UpgradeBeaconProxy} from "@nomad-xyz/contracts-core/contracts/upgrade/UpgradeBeaconProxy.sol";
+
+contract DeployAccountant is Script, Config {
+    // entrypoint
+    function deploy(string calldata configFile, string calldata domain) public {
+        __Config_initialize(configFile);
```
[the remark this answers is not in the document]
Comment (anna-carroll) [reply packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol:15]: good call, fixed https://github.com/nomad-xyz/monorepo/pull/501/commits/b11f24438ef857ee786419f987f815ba8ae0b668

File: packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        //       because the initialize function has no parameters
+        //       so `initialize` will not be called automatically on deployment
+        AllowListNFTRecoveryAccountant(address(proxy)).initialize();
+        // transfer ownership of proxy
+        AllowListNFTRecoveryAccountant(address(proxy)).transferOwnership(
+            accountantOwner
+        );
+    }
+
+    function write() internal {
```
Comment (anna-carroll) [inline packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol:56]: `vm.writeJSON` wen 🥰

File: packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        string memory json = '{\n   "accountant": {\n';
```
Comment (prestwich) [inline packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol:57]: should we make a `JsonWriter` library contract?

Comment (odyslam) [reply packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol:57]: https://github.com/foundry-rs/foundry/pull/3595

File: packages/contracts-ops/contracts/Config.sol [hunk 6]
```diff
@@ -182,6 +183,11 @@ abstract contract Config {
             abi.decode(loadBridgeAttribute(domain, "deployHeight"), (uint256));
     }
 
+    function bridgeRouter(string memory domain) public onlyInitialized returns (BridgeRouter) {
```
[the remark this answers is not in the document]
Comment (anna-carroll) [reply packages/contracts-ops/contracts/Config.sol:186]: fixed INomadProtocol PR. closing here, will rebase this on that when it's merged

File: packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        // beacon
+        json = string(
+            abi.encodePacked(
+                json,
+                '"      beacon": ',
+                '"',
+                vm.toString(address(beacon)),
+                '",\n'
```
[the remark this answers is not in the document]
Comment (anna-carroll) [reply packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol]: i'm going to delete the json writing func until we merge the library https://github.com/nomad-xyz/monorepo/pull/501/commits/6c94109e84b269564666e705aa85dca769c48a6f

File: packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol (hunk 2 above)
Comment (prestwich) [reply packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol:31]: okay, seems better than anything else we're doing. I'll prep adding them to the config rust

Comment (anna-carroll) [reply packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol:31]: danke schon

File: packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        implementation.initialize();
+        // deploy UpgradeBeacon (with UpgradeBeaconController as owner)
+        beacon = new UpgradeBeacon(
+            address(implementation),
+            address(upgradeBeaconController(_domain))
+        );
+        // deploy UpgradeBeaconProxy
+        proxy = new UpgradeBeaconProxy(
+            address(beacon),
+            "0x"
```
Comment (prestwich) [inline packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol:57]: I think you mean `""`

**Review by prestwich — CHANGES_REQUESTED** [review]
initialization currently broken. passes bytestring "0x" instead of empty bytestring

File: packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol (hunk 8 above)
Comment (anna-carroll) [reply packages/contracts-ops/contracts/scripts/DeployAccountant.s.sol]: https://github.com/nomad-xyz/monorepo/pull/501/commits/1a99637a989ff0a81c1ffc913aff1be2a5ff29af

**Review by prestwich — APPROVED** [review]

**Pull request merged**